compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -682,16 +682,304 @@ abstract class GenJSCode extends plugins.PluginComponent
             "repeated parameters.")
       }
 
-      // Implementation restriction
-      for (tree <- secondaryCtorTrees) {
-        reporter.error(tree.pos,
-            "Implementation restriction: Scala.js-defined JS classes cannot " +
-            "have secondary constructors")
+      withNewLocalNameScope {
+        val infoBuilder = new MethodInfoBuilder()
+        infoBuilder.setEncodedName("constructor")
+        infoBuilder.setIsExported(true)
+        infoBuilder.setIsStatic(false)
+
+        val infoBuilderOpt = Some(infoBuilder)
+
+        val primaryCtor =
+          genMethodWithInfoBuilder(primaryCtorTree, infoBuilderOpt).get._1
+        val secondaryCtors = secondaryCtorTrees.map { tree =>
+          genMethodWithInfoBuilder(tree, infoBuilderOpt).get._1
+        }
+
+        val dispatch =
+          genJSConstructorExport(constructorTrees.map(_.symbol), infoBuilder)
+
+        val isConstructor =
+          secondaryCtors.map(_.name.name).toSet + primaryCtor.name.name
+
+        val constructorGraph =
+          mkConstructorGraph(primaryCtor, secondaryCtors, isConstructor)
+
+        val overloadIdent = js.Ident("overload")
```

Comment:
This needs to be a `freshIdent`, otherwise it will collide with a local variable named `overload` in one of the constructors.